I have a problem with this fix: it produces dubious C code.  At least according to gcc and to clang, an `extern` declaration with an initializer is a likely error in C:
```
● cat foo.c
extern const char * x[2] = { "a", "b" };
● clang -std=c11 -Wall -c foo.c
foo.c:1:21: warning: 'extern' variable has an initializer [-Wextern-initializer]
    1 | extern const char * x[2] = { "a", "b" };
      |                     ^
1 warning generated.
● gcc -std=c11 -Wall -c foo.c
foo.c:1:21: warning: ‘x’ initialized and declared ‘extern’
    1 | extern const char * x[2] = { "a", "b" };
      |                     ^
```

--- a/Changes
+++ b/Changes
@@ -514,6 +514,11 @@ Working version
 
 ### Bug fixes:
 
+- # 12791: `extern` is applied to definitions of `caml_builtin_cprim`
+    and `caml_names_of_builtin_cprim` when linking bytecode '-custom'
+    executables with a C++ linker
+  (Shayne Fletcher, review by Antonin Décimo and Xavier Leroy)
+
 - #12712, #12742: fix an assertion boundary case in `caml_reset_young_limit`
   (Jan Midtgaard, review by Guillaume Munch-Maccagnoni and KC Sivaramakrishnan)
 

--- a/bytecomp/symtable.ml
+++ b/bytecomp/symtable.ml
@@ -177,11 +177,17 @@ let output_primitive_table outchan =
     fprintf outchan "extern value %s(void);\n" prim.(i)
   done;
   fprintf outchan "typedef value (*c_primitive)(void);\n";
+  fprintf outchan "#if defined __cplusplus\n";
+  fprintf outchan "extern\n";
+  fprintf outchan "#endif\n";
   fprintf outchan "const c_primitive caml_builtin_cprim[] = {\n";
   for i = 0 to Array.length prim - 1 do
     fprintf outchan "  %s,\n" prim.(i)
   done;
   fprintf outchan "  0 };\n";
+  fprintf outchan "#if defined __cplusplus\n";
+  fprintf outchan "extern\n";
+  fprintf outchan "#endif\n";
   fprintf outchan "const char * const caml_names_of_builtin_cprim[] = {\n";
   for i = 0 to Array.length prim - 1 do
     fprintf outchan "  \"%s\",\n" prim.(i)

